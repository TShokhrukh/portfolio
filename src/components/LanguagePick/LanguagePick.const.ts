import { cn } from '@bem-react/classname'

export const cnLanguagePick = cn('LanguagePick')
