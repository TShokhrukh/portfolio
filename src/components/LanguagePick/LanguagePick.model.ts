import { observable, makeObservable, action } from 'mobx'
import type { ILanguagePickModel } from './__types__'

export class LanguagePickModel implements ILanguagePickModel {
  public _supported: string[]
  @observable public current: string

  constructor (
    current: string,
    supported: string[]
  ) {
    this.current = current
    this._supported = supported

    makeObservable(this)
  }

  get supported (): string[] {
    return this._supported.filter(l => l !== this.current)
  }

  @action
  setLocal (language: string) {
    this.current = language
  }
}
