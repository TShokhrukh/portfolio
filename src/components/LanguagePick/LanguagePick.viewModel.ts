import { makeObservable, observable, action } from 'mobx'
import type { ILanguagePickViewModel, ILanguagePickModel } from './__types__'
import type { IPopperModel } from '../Popper/__types__'

export class LanguagePickViewModel implements ILanguagePickViewModel {
  @observable.ref private language: ILanguagePickModel
  @observable.ref private popper: IPopperModel

  constructor (
    languageModel: ILanguagePickModel,
    popperModel: IPopperModel
  ) {
    this.language = languageModel
    this.popper = popperModel

    makeObservable(this)
    this.open = this.open.bind(this)
    this.close = this.close.bind(this)
    this.toggle = this.toggle.bind(this)
  }

  get supported (): string[] {
    return this.language.supported
  }

  get current (): string {
    return this.language.current
  }

  get isOpen (): boolean {
    return this.popper.isOpen
  }

  @action
  setLocal (language: string): void {
    this.language.setLocal(language)
    this.close()
  }

  @action
  open (): void {
    this.popper.setIsOpen(true)
  }

  @action
  close (): void {
    this.popper.setIsOpen(false)
  }

  @action
  toggle (): void {
    this.popper.setIsOpen(!this.isOpen)
  }
}
