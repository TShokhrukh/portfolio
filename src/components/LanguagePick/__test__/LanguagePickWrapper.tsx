import React, { ReactNode } from 'react'
import { TranslationsProvider } from '@eo-locale/react'

import type { Locale } from '@eo-locale/core'

interface IWrapperProps {
  children: ReactNode
}

const locales: Locale[] = [
  {
    language: 'en',
    messages: {
      language: {
        en: 'en',
        ru: 'ru',
        uz: 'uz'
      }
    }
  }
]

export const LanguagePickWrapper: React.FC<IWrapperProps> = ({ children }) => {
  return (
    <TranslationsProvider language='en' locales={locales} onError={() => null}>
      {children}
    </TranslationsProvider>
  )
}
