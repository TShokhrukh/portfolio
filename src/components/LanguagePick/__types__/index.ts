export * from './LanguagePick.types'
export * from './LanguagePick.model.types'
export * from './LanguagePick.viewModel.types'
