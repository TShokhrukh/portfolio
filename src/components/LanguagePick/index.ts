export * from './build'
export * from './__types__'
export * from './LanguagePick.viewModel'
export * from './LanguagePick.model'
export { LanguagePick as LanguagePickBase } from './LanguagePick'
