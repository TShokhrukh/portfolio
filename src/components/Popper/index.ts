export * from './build'
export * from './__types__'
export * from './Popper.viewModel'
export * from './Popper.model'
export { Popper as PopperBase } from './Popper'
