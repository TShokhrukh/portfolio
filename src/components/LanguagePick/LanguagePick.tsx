import React, { useRef } from 'react'
import { Observer } from 'mobx-react'
import { Text } from '@eo-locale/react'
import { Button } from '../Button/Button'
import { Popper } from '../Popper/Popper'
import { cnLanguagePick } from './LanguagePick.const'

import type { ILanguagePickProps } from './__types__'

export const LanguagePick: React.FC<ILanguagePickProps> = ({
  id,
  viewModel,
  className,
  ...props
}) => {
  const anchorEl = useRef<HTMLElement>(null)
  return (
    <Popper id={id} viewModel={viewModel} anchorEl={anchorEl}
      target={(targetProps) => (
        <Button {...props} {...targetProps} innerRef={anchorEl} className={cnLanguagePick({}, [className])}>
          <Observer>
            {() => <Text id={`language.${viewModel.current}`} />}
          </Observer>
        </Button>
      )}>
      { viewModel.supported.map(lang => (
        <Button key={lang} onClick={() => viewModel.setLocal(lang)}>
          <Text id={`language.${lang}`} />
        </Button>
      )) }
    </Popper>
  )
}

LanguagePick.defaultProps = {
  id: cnLanguagePick()
}
