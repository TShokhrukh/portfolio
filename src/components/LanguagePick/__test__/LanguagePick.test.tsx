/* global test, expect, describe, beforeEach */
import React from 'react'
import { act, render, screen } from '@testing-library/react'
// import { act } from 'react-dom/test-utils'
import { LanguagePick } from '../LanguagePick'
import { LanguagePickModel } from '../LanguagePick.model'
import { LanguagePickViewModel } from '../LanguagePick.viewModel'
import { PopperModel } from '../../Popper/Popper.model'
import { LanguagePickWrapper } from './LanguagePickWrapper'

import type { ILanguagePickModel, ILanguagePickViewModel } from '../__types__'
import type { IPopperModel } from '../../Popper/__types__'

describe('LanguagePick', () => {
  let model: ILanguagePickModel
  let popperModel: IPopperModel
  let viewModel: ILanguagePickViewModel

  beforeEach(() => {
    model = new LanguagePickModel('en', ['en', 'ru', 'uz'])
    popperModel = new PopperModel()
    viewModel = new LanguagePickViewModel(model, popperModel)
  })

  test('should return the component template', () => {
    render(
      <LanguagePickWrapper>
        <LanguagePick viewModel={viewModel} id='test_id' className='test_class' data-testid='language_pick' />
      </LanguagePickWrapper>
    )

    expect(screen.getByTestId('language_pick')).toMatchSnapshot()
  })

  test('should render content if viewModel.isOpen property is "true"', async () => {
    render(
      <LanguagePickWrapper>
        <LanguagePick viewModel={viewModel} data-testid='language_pick' />
      </LanguagePickWrapper>
    )

    expect(screen.getByTestId('language_pick').textContent).toBe('en')
    expect(screen.queryByText('uz')).not.toBeInTheDocument()
    expect(screen.queryByText('ru')).not.toBeInTheDocument()

    await act(async () => {
      await viewModel.open()
    })

    expect(screen.queryAllByText('en').length).toBeLessThanOrEqual(1)
    expect(screen.queryByText('uz')).toBeInTheDocument()
    expect(screen.queryByText('ru')).toBeInTheDocument()
  })
})
