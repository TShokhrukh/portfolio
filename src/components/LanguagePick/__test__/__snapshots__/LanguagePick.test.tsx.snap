// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LanguagePick must return the component template 1`] = `
<button
  aria-describedby="test_id"
  class="Button LanguagePick test_class"
  data-testid="language_pick"
  type="button"
>
  en
</button>
`;
